src/BlockNLPModels.jl

```diff
@@ -0,0 +1,79 @@
+module BlockNLPModels
+
+using NLPModels
+using LinearAlgebra
+
+export AbstractBlockNLP
+export BlockNLPModel
+export DualizedNLPblockModel
+export FullSpaceModel
+
+abstract type AbstractBlockNLP end
+
+mutable struct BlockNLPModel{T<:AbstractNLPModel} <: AbstractBlockNLP
+  blocks::Vector{T}
+  linkconstraints::Vector{Float64}
+end
+
+mutable struct DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}
+  meta::NLPModelMeta{T, S}
+  counters::Counters
+  problem_block::AbstractNLPModel
+  λ::AbstractVector # dual variables
+  a::AbstractVector # linking constraint coeff
+end
+
+function DualizedNLPblockModel(nlp::AbstractNLPModel, λ::AbstractVector, a::AbstractVector)
+  meta = nlp.meta
+  return DualizedNLPblockModel(meta, Counters(), nlp, λ, a)
+end
+
+function NLPModels.obj(nlp::DualizedNLPblockModel, x::AbstractVector)
+  n = nlp.problem_block.meta.nvar
+  d(x) = dot(nlp.λ, dot(nlp.a, x))
+  return obj(nlp.problem_block, x) + d(x)
+end
+
+function NLPModels.grad!(nlp::DualizedNLPblockModel, x::AbstractVector, g::AbstractVector)
+  grad!(nlp.problem_block, x, g)
+  g .+= dot(nlp.λ, nlp.a)
+  return g
+end
```

Comment:
Same question as in `obj`. Perhaps we can also provide a "default" implementation which does autodiff on the `obj` function? What do you think @frapac ?

Reply:
I will let @frapac respond to this, but I just went with Sungho's and Francois's idea of BlockNLPModels.jl being a thin wrapper.

Reply:
I agree we should keep BlockNLPModels as thin as possible. If the users are interested to use AutoDiff, they can use ADNLPModels.jl.